src/emmy/numerical/ode.cljc

```diff
@@ -186,19 +186,24 @@
   (let [evaluation-time    (us/stopwatch :started? false)
         evaluation-count   (atom 0)
         flat-initial-state (flatten initial-state)
+        primitive-params   (double-array derivative-args)
         derivative-fn      (if compile?
-                             (let [f' (c/compile-state-fn state-derivative derivative-args initial-state)]
-                               (fn [y]
-                                 (f' y (or derivative-args []))))
+                             (c/compile-state-fn state-derivative derivative-args initial-state
+                                                 {:calling-convention :primitive
+                                                  :generic-params? false})
                              (do (log/warn "Not compiling function for ODE analysis")
-                                 (let [d:dt (apply state-derivative derivative-args)
-                                       array->state #(struct/unflatten % initial-state)]
-                                   (comp d:dt array->state))))
+                                 (let [d:dt (apply state-derivative derivative-args)]
+                                   (fn [ys yps _]
+                                     (flatten-into-primitive-array yps (d:dt (struct/unflatten ys initial-state))))
+                                   )))
         equations        (fn [_ y out]
+                           ;; TODO: should we consider allowing an option to add a dummy
```

Comment:
The double-decker state derivative is annoying to deal with in this case since (in the SICM-book-compatible cases) we want to instrument the inner function that is returned from the application to `derivative-args`. So I'm going to take it out for now--as you say the customer can arrange for it themselves